### PR TITLE
Robustness improvements

### PR DESCRIPTION
**Pull request by indygreg**
I took the time to write detailed commit messages for each commit. This series partially addresses some of the issues in bug 1181783.

### CODE REVIEW

File: src/cli/clone.js [hunk 1]
```diff
@@ -45,10 +45,15 @@ export default async function main(config, argv) {
...
-    let vcsConfig = await detect(args.url);
-    let vcs = require('../vcs/' + vcsConfig.type);
+  let vcsConfig = await detect(args.url);
+  let vcs = require('../vcs/' + vcsConfig.type);
+
+  // If we used a cache, pull from latest upstream so we have latest
+  // commits. Else, fall back to a full clone. The end state should be
+  // reasonable consistent.
+  if (usedCache) {
+    await vcs.pull(config, args.dest, args.url);
```
Comment (catlee) [inline src/cli/clone.js]: are the dest/url arguments really opposite in vcs.pull really opposite of vcs.clone?

**Comment by jonasfj** [comment]
@gregarndt, I see no problems here what do you think...

Note, when deploying this we really should add some checks in bash to make sure we got the right revision, we previously had issues with tc-vcs not exiting non-zero on failure to update.

**Comment by gregarndt** [comment]
From what little I've looked at with tc-vcs, this looks good to me.  I can understanding the reasoning behind only having one retry.

**Comment by jonasfj** [comment]
One retry is a bit fragile, slightly related I proposed exponential back-off in PR #19.

**Comment by indygreg** [comment]
I also prefer exponential backoff with jitter than the quick and dirty change here.

**Comment by indygreg** [comment]
Also, the change to add a pull after cached unarchive may not be necessary if the caches are updated every day. This PR is effectively working around symptoms of having repo caches that are several days out of date.

**Pull request merged**